### PR TITLE
A bunch of refactorings

### PR DESCRIPTION
**Pull request by adamcooke**
This PR introdues a bunch of refactorings and will become Authie 4.0.

* switch to rspec
* remove impersonation (for now)
* add session/controller proxy to tidy session model
* tidy config
* remove current_user=
* other bits and bobs

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -8,33 +8,180 @@ PATH
 GEM
   remote: https://rubygems.org/
   specs:
-    activemodel (6.1.5.1)
-      activesupport (= 6.1.5.1)
-    activerecord (6.1.5.1)
-      activemodel (= 6.1.5.1)
-      activesupport (= 6.1.5.1)
-    activesupport (6.1.5.1)
+    actioncable (7.0.2.4)
```
Comment (jimeh) [inline Gemfile.lock:11]: I'd vote for removing the `Gemfile.lock` from the repo, as this is a gem, so we shouldn't lock gems to any specific version.

File: gemfiles/activerecord_6.1.gemfile.lock [hunk 2]
```diff
@@ -2,12 +2,26 @@ PATH
   remote: ..
   specs:
     authie (0.0.0.dev)
+      actionpack (>= 5.0, < 8.0)
```
Comment (jimeh) [inline gemfiles/activerecord_6.1.gemfile.lock:5]: I believe all the `activerecord*.` files here should be removed, as the Appraisal groups have been renamed to `rails*`.

File: gemfiles/rails_6.1.gemfile.lock [hunk 3]
```diff
@@ -0,0 +1,232 @@
+PATH
```
Comment (jimeh) [inline gemfiles/rails_6.1.gemfile.lock:1]: Similarly to `Gemfile.lock`, we probably shouldn't commit the `*.lock` files, so as to ensure tests run against latest versions of deps that meet specified constraints, so it matches up with whatever dependency versions get used when someone adds authie to their project.

File: lib/authie/session.rb [hunk 4]
```diff
@@ -1,338 +1,264 @@
...
+    delegate :set, to: :session
+    delegate :temporary_token, to: :session
+    delegate :token_hash, to: :session
+    delegate :two_factored_at, to: :session
+    delegate :two_factored_ip, to: :session
+    delegate :two_factored?, to: :session
+    delegate :update, to: :session
+    delegate :update!, to: :session
+    delegate :user_agent, to: :session
+    delegate :user, to: :session
```
Comment (jimeh) [inline lib/authie/session.rb:262]: Correct me if I'm wrong, but I believe the `delegate` method here comes from ActiveSupport, meaning this class will fail to parse unless ActiveSupport/Rails has already been loaded.

I'd suggest adding `activesupport` as a runtime gem dependency, and then adding this require call to the top of the file to ensure this works in isolation:

```ruby
require 'active_support/core_ext/module/delegation'
```

File: lib/authie/session_model.rb [hunk 5]
```diff
@@ -0,0 +1,140 @@
+# frozen_string_literal: true
+
+require 'secure_random_string'
+require 'authie/config'
+
+module Authie
+  class SessionModel < ActiveRecord::Base
```
Comment (jimeh) [inline lib/authie/session_model.rb:7]: I'd suggest requiring `active_record/base` in this file to ensure the `SessionModel` class can be parsed if authie is loaded before/without ActiveRecord/Rails.

**Review by jimeh — CHANGES_REQUESTED** [review]

File: lib/authie/session_model.rb (hunk 5 above)
Comment (adamcooke) [reply lib/authie/session_model.rb:7]: Done

**Pull request merged**